### PR TITLE
test: use random free port for end2end tests

### LINKED ISSUE #786: chore: use random port for test server
See https://github.com/influxdata/influxdb_iox/issues/745#issuecomment-774092141

### LINKED ISSUE #2403: Flaky Test: Server ID Conflict
When running tests locally on current main, I'm consistently seeing these two tests failing

```
failures:
    end_to_end_cases::management_api::test_wipe_preserved_catalog
    end_to_end_cases::write_pb::test_write_pb
```

The errors appear to be test races, but I've not looked into it yet

test_wipe_preserved_catalog

```
Server already has an ID (32); possibly a stray/orphan server from another test run.
```

test_write_pb

```
Server already has an ID (42); possibly a stray/orphan server from another test run.
```

### PR DESCRIPTION
**Pull request by crepererum**
Fixes #786.
Fixes #2403.

### CODE REVIEW

File: tests/common/server_fixture.rs [hunk 1]
```diff
@@ -47,24 +40,34 @@ impl BindAddresses {
...
+
+    fn get_free_port() -> SocketAddrV4 {
+        let mut rng = rand::thread_rng();
+        let ip = std::net::Ipv4Addr::new(127, 0, 0, 1);
+
+        loop {
+            let port = rng.gen_range(8000..9000);
+            let addr = SocketAddrV4::new(ip, port);
+            if std::net::TcpListener::bind(addr).is_ok() {
+                return addr;
```
Comment (shepmaster) [inline tests/common/server_fixture.rs:52]: Does anything prevent another process from grabbing this port between here and when the server actually starts?

File: tests/common/server_fixture.rs [hunk 2]
```diff
@@ -47,24 +40,34 @@ impl BindAddresses {
...
 impl Default for BindAddresses {
     /// return a new port assignment suitable for this test's use
     fn default() -> Self {
-        let http_port = NEXT_PORT.fetch_add(1, SeqCst);
-        let grpc_port = NEXT_PORT.fetch_add(1, SeqCst);
-
-        let http_bind_addr = format!("127.0.0.1:{}", http_port);
-        let grpc_bind_addr = format!("127.0.0.1:{}", grpc_port);
+        let http_addr = Self::get_free_port();
+        let grpc_addr = Self::get_free_port();
```
Comment (shepmaster) [inline tests/common/server_fixture.rs:62]: It’s possible that these two ports are the same, due to randomness, correct?

or even two parallel tests grabbing the same random value(s)?

File: tests/common/server_fixture.rs [hunk 3]
```diff
@@ -47,24 +40,34 @@ impl BindAddresses {
     pub fn grpc_bind_addr(&self) -> &str {
         &self.grpc_bind_addr
     }
+
+    fn get_free_port() -> SocketAddrV4 {
+        let mut rng = rand::thread_rng();
+        let ip = std::net::Ipv4Addr::new(127, 0, 0, 1);
+
+        loop {
+            let port = rng.gen_range(8000..9000);
```
Comment (shepmaster) [inline tests/common/server_fixture.rs:49]: Isn’t there a dedicated range for ephemeral ports?

File: tests/common/server_fixture.rs (hunk 1 above)
Comment (crepererum) [reply tests/common/server_fixture.rs:52]: No, but I don't know if there's anything at all we can do about it (except for making the server more complicated and support passing an FD around).

File: tests/common/server_fixture.rs (hunk 2 above)
Comment (crepererum) [reply tests/common/server_fixture.rs:62]: Uhh, good catch! We should probably hold the first port while scanning for the second.

File: tests/common/server_fixture.rs (hunk 3 above)
Comment (crepererum) [reply tests/common/server_fixture.rs:49]: Does this count as ephemeral though? It's not really the client-side here, but we're looking for a server port that we're potentially connect to multiple times.

File: tests/common/server_fixture.rs (hunk 2 above)
Comment (crepererum) [reply tests/common/server_fixture.rs]: fixed

**Review by NGA-TRAN — APPROVED** [review]
LGTM

File: tests/common/server_fixture.rs [hunk 4]
```diff
@@ -41,25 +42,35 @@ impl BindAddresses {
...
         let mut rng = rand::thread_rng();
         let ip = std::net::Ipv4Addr::new(127, 0, 0, 1);
 
-        loop {
+        let mut results = BTreeSet::new();
+
+        while results.len() < n {
             let port = rng.gen_range(8000..9000);
             let addr = SocketAddrV4::new(ip, port);
+            if results.contains(&addr) {
```
Comment (shepmaster) [inline tests/common/server_fixture.rs:54]: The check is redundant since you have a set and the loop is on the length of the set.

Comment (crepererum) [reply tests/common/server_fixture.rs]: fixed

**Comment by crepererum** [comment]
I need another round on this tomorrow. This doesn't work w/ multi-threaded tests (which we use by default). So we need a hybrid between the new and the old solution (aka using some process-global explicit or implicit set but also check if the port is usable).

**Comment by shepmaster** [comment]
> a hybrid between the new and the old solution

I was going to suggest something like that. The old version with the monotonically increasing ports that skips a port if it can detect that the port is currently in use. We don't actually _desire_ randomness for this.

**Pull request merged**